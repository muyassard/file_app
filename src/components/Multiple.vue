<script setup lang="ts">
  import { ref } from "vue";

  import { CreateImage } from "@/modules/api";
  import type { Image } from "@/modules/type";

  import defaultImage from "@/images/defaultImage.png";

  import { message } from "antd";
  import { Status } from "@/util/status";
  import { getFileType } from "@/util/fileType";

  interface Props {
    maxSize: number;
    limit: number;
    type: string;
  }
  const props = defineProps<Props>();

  const multiData = ref<Image[]>([]);
  const isLoading = ref(false);

  const change = async (e: Event) => {
    let files = (e.target as HTMLInputElement).files;

    if (!files?.length) return;

    let len = files.length;

    if (files.length > props.limit) {
      len = props.limit;
      message.error("you are uploading a lot of files");
    }

    for (let i = 0; i < len; i++) {
      if (!files[i].type.includes(props.type)) {
        message.error("you cannot upload such a file");
      } else {
        multiData.value.push({
          photo: getFileType(files[i]),
          file: files[i],
          status: "",
          id: files[i].lastModified,
        });
      }
      if (multiData.value.length >= props.limit) {
        multiData.value.length = props.limit;
      }
    }
    console.log(multiData.value);
  };

  const upload = async () => {
    if (multiData.value.length > 0) {
      try {
        isLoading.value = true;

        multiData.value.forEach(async (multi) => {
          if (multi.status !== Status.success && multi.status !== Status.failed) {
            if (multi.file?.size! > props.maxSize) {
              multi.status = Status.failed;
              message.error("This image size is large");
            } else {
              await CreateImage(multi.file!);
              multi.status = Status.success;
            }
          }
        });
      } catch (error) {
        console.log(error);
      } finally {
        isLoading.value = false;
      }
    } else {
      message.error("you have not selected a file yet");
    }
  };

  const remove = (id: number) => {
    const removeElm = multiData.value.filter((file: Image) => file.id !== id);
    multiData.value = removeElm;
  };

  defineExpose({
    change,
    upload,
    remove,
  });
</script>

<template>
  <div class="flex items-end gap-2">
    <div class="grid gap-3 w-[200px]">
      <div class="">Multi</div>
      <label
        :style="{
          backgroundImage: `url(${defaultImage})`,
        }"
        class="image border"
      >
        <input
          data-type="multipleInput"
          @change="change"
          type="file"
          class="hidden"
          multiple
        />
      </label>
      <b class="">you can only upload a {{ props.type }} file</b>

      <button
        data-type="multipleUploadButton"
        @click="upload"
        :disabled="isLoading"
        class="text-white bg-blue-500 w-[200px]"
      >
        {{ isLoading ? "uploading..." : "upload" }}
      </button>
    </div>
    <div class="flex flex-wrap gap-2">
      <div class="relative" v-for="multi in multiData" :key="multi.id">
        <img class="w-[200px] h-[200px] object-cover" :src="multi.photo" alt="" />
        <img class="absolute top-0 left-1 w-[30px]" :src="multi.status" />
        <div class="mb-2">{{ multi.file?.name }}</div>
        <button class="bg-blue-500 text-white w-[200px]" @click="() => remove(multi.id)">
          Remove
        </button>
      </div>
    </div>
  </div>
</template>

<style scoped>
  .image {
    width: 200px;
    height: 200px;
    background-position: center;
    background-size: cover;
  }
</style>
