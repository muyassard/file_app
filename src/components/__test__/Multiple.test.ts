import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils';

import { nextTick } from 'vue';
import Multiple from '../Multiple.vue';

describe('Multiple', () => {
    it('adds files that do not exceed the limit to multiData', async () => {
        const wrapper = mount(Multiple, {
            props: {
                maxSize: 1000000,
                limit: 3,
                type: 'zip',
            },
        });


        const inputElement = wrapper.find('input[type="file"]').element as HTMLInputElement;
        const mockFileList = Object.create(inputElement.files);

        const file1 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        const file2 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        mockFileList[0] = file1;
        mockFileList[1] = file2;

        Object.defineProperty(mockFileList, 'length', { value: 2 });
        (wrapper.getCurrentComponent().exposed as unknown as any).change({
            target: { files: mockFileList },
        });


        await nextTick();

        const multiData = wrapper.find("multiData")

        expect(wrapper.vm.multiData.length).toBe(2);
    });

    it('limits the number of files added to multiData when exceeding the limit', async () => {
        const wrapper = mount(Multiple, {
            props: {
                maxSize: 1000000,
                limit: 2,
                type: 'zip',
            },
        });

        const inputElement = wrapper.find('input[type="file"]').element as HTMLInputElement;
        const mockFileList = Object.create(inputElement.files);

        const file1 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        const file2 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        const file3 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        mockFileList[0] = file1;
        mockFileList[1] = file2;
        mockFileList[2] = file3;

        Object.defineProperty(mockFileList, 'length', { value: 3 });
        (wrapper.getCurrentComponent().exposed as unknown as any).change({
            target: { files: mockFileList },
        });


        await nextTick();

        expect(wrapper.vm.multiData.length).toBe(2);
    });

    it('if another type of file arrives, it will not be added to multiData', async () => {
        const wrapper = mount(Multiple, {
            props: {
                maxSize: 1000000,
                limit: 2,
                type: 'pdf',
            },
        });

        const inputElement = wrapper.find('input[type="file"]').element as HTMLInputElement;
        const mockFileList = Object.create(inputElement.files);


        const file1 = new File(['2777'], 'foo.pdf', {
            type: 'application/pdf',
        });

        const file2 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        const file3 = new File(['2777'], 'foo.zip', {
            type: 'application/zip',
        });
        mockFileList[0] = file1;
        mockFileList[1] = file2;
        mockFileList[2] = file3;




        Object.defineProperty(mockFileList, 'length', { value: 3 });
        (wrapper.getCurrentComponent().exposed as unknown as any).change({
            target: { files: mockFileList },
        });

        await nextTick();

        expect(wrapper.vm.multiData.length).toBe(1);
    });





});
