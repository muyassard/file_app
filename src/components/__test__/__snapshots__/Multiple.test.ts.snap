// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Single > should render File input 1`] = `"<input data-v-91122bb6="" type="file" class="hidden">"`;
